tenders/common/tap_attach.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2015-2019 Contributors as noted in the AUTHORS file
+ *
+ * This file is part of Solo5, a sandboxed execution environment.
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * tap_attach.c: Common functions for attaching to TAP interfaces.
+ */
+
+#include <assert.h>
+#include <err.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <ifaddrs.h>
+#include <limits.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#if defined(__linux__)
+
+/*
+ * Linux TAP device specific.
+ */
+#include <sys/socket.h>
+#include <linux/if.h>
+#include <linux/if_tun.h>
+
+#elif defined(__FreeBSD__)
+
+#include <net/if.h>
+
+#elif defined(__OpenBSD__)
+
+#include <sys/socket.h>
+#include <net/if.h>
+
+#else /* !__linux__ && !__FreeBSD__ && !__OpenBSD__ */
+
+#error Unsupported target
+
+#endif
+
+int tap_attach(const char *ifname)
+{
+    int fd;
+
+    /*
+     * Syntax @<number> indicates a pre-existing open fd, so just pass it
+     * through if the supplied <number> is in range and O_NONBLOCK can be set.
+     */
+    if (ifname[0] == '@') {
+	char *endp;
+	long int maybe_fd = strtol(&ifname[1], &endp, 10);
+	if (*endp != 0 /* Invalid character at (*endp)? */
+		|| endp == &ifname[1] /* Empty string? */)
+	    errno = EINVAL;
+	else if (maybe_fd < 0 || maybe_fd > INT_MAX)
+	    errno = ERANGE;
+	if (errno)
+	    err(1, "tap_attach");
+        
+	fd = (int)maybe_fd;
+        if (fcntl(fd, F_SETFL, O_NONBLOCK) == -1)
+            return -1;
+
+        return fd;
+    }
+
+    /*
+     * Verify that the interface exists and is up and running. If we don't do
+     * this then we get "create on open" behaviour on most systems which is not
+     * what we want.
+     */
+    struct ifaddrs *ifa, *ifp;
+    int found = 0;
+    int up = 0;
+
+    if (getifaddrs(&ifa) == -1)
+        return -1;
+    ifp = ifa;
+    while (ifp) {
+        if (strcmp(ifp->ifa_name, ifname) == 0) {
+            found = 1;
+            up = ifp->ifa_flags & (IFF_UP | IFF_RUNNING);
+            break;
+        }
+        ifp = ifp->ifa_next;
+    }
+    freeifaddrs(ifa);
+    if (!found) {
+        errno = ENOENT;
+        return -1;
+    }
+
+#if defined(__linux__)
+
+    if (!up) {
```

Comment:
> is it intentional that the FreeBSD version doesn't check the status of up?

Probably (see below).

> up is defined as ifp->ifa_flags & (IFF_UP | IFF_RUNNING), so these checks determine if either one is set, but it doesn't require both. Is that intentional?

Some `tap` implementations (I forget which Linux/BSD) will return an immediate error if you try to write packets to a `tap` interface which is not "configured" and "up" (for some version of, again, it differs between systems) on the host side. AFAIR this code is trying to prevent the caller/user from attaching to such an interface in the first place, to prevent confusing errors later down the line.

I agree that the intent is not clear here, and the addition of different backends (for the BSDs) has complicated things. If we want to clean this up, I'd have to re-test manually on the different systems to see what the exact behaviour is and if we can unify it.

However, I think this one is a case of "if it ain't broke, don't fix it", so I'd probably leave it alone for now unless there's some case that is definitely broken.

Reply:
this is a good point -- I remember myself debugging in Marrakesh when setting up the router exactly this issue (tap device was not up, the unikernel received packets but was unable to send any). The underlying issue was that while I had `sysctl net.link.tap.up_on_open=1` in `/etc/sysctl.conf`, `tap` was compiled as a kernel module, and only loaded after `sysctl.conf` was evaluated -- leading to `net.link.tap.up_on_open` being 0.

the `getifaddrs` at the near top of this function is called before `open`, thus on FreeBSD `up` is 0. a fix which I locally tested is to call `getifaddrs` again after `open` and look whether the `IFF_UP` flag is present.

i wonder whether there are any (valid) uses where the `tap` interface is down (and should be -- maybe for a passive listener unikernels). I don't know anyone using tap interface and leaving `net.link.tap.up_on_open=0`.

so, if you'd accept another call to getifaddrs in the FreeBSD-specific part of `tap_attach`, let me know and I could PR that.

Reply:
@hannesm Ah, that explains the missing `if (!up) {`.

We could try to set it `IFF_UP` with something like this, although automatically setting an interface to `UP` seems a little bit yolo(?):
```c
/* may need some of these in the FreeBSD include block above: */
#include <sys/param.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <net/if.h>
#include <net/if_var.h>
#include <net/if_types.h>
#include <sockio.h>


/* in the FreeBSD block: */
if (!up) {
    struct ifreq ifr;
    memset(&ifr, 0, sizeof(ifr));
    strncpy(ifr.ifr_name, ifname, IFNAMSIZ - 1);
    ifr.ifr_flags = IFF_UP; /* TODO do we need to read the ifr_flags first and |= IFF_UP ?*/
    if (ioctl(fd, SIOCSIFFLAGS, (void *)&ifr) == -1) {
        err = errno;
        close(fd);
        errno = err;
        return -1;
    }
}
```

Could potentially qualify the decision by checking that `net.link.tap.up_on_open == 0` with something like:
```c
if (!up) {
    int up_on_open = 0;
    int err = sysctlbyname("net.link.tap.up_on_open",
            &up_on_open, 0, NULL, 0);
    if (err) {
        /* yeah well */
    }
    if (!up_on_open) {
        /* issue ioctl to set it IFF_UP */
    }
}
```

As for the bigger picture, couldn't one imagine a firewall unikernel that might want to attach to an interface regardless of up/down status so that it was ready to start handling packets when it comes up? I'm not so familiar with network interfaces, so I don't know what the use-cases are for having downed interfaces.